scipy/sparse/_construct.py

```diff
@@ -481,9 +481,9 @@ def kronsum(A, B, format=None):
     return (L+R).asformat(format)  # since L + R is not always same format
 
 
-def _compressed_sparse_stack(blocks, axis):
+def _compressed_sparse_stack(blocks, axis, container):
```

Comment:
This PR doesn't change the `idx_dtype` behavior.
But also, I don't think we need to add `indices` to `get_index_dtype` because `indptr` will indicate the current type and the max value needed. So I think this is OK as is...

Reply:
This PR enables concatenation for the array classes, which have different index behavior that matrix classes. So there will be no existing tests for index dtype behavior.

I believe `indptr` will only indicate the correct index type after certain operations that check for this. `indptr` and `indices` sharing a `dtype` is not an invariant of these classes:

```python
csr = sparse.random(10, 10, density=0.25, format="csr")
csr.indices = csr.indices.astype(np.int64)
(csr.indices.dtype, csr.indptr.dtype)
# (dtype('int64'), dtype('int32'))
```

---

Here's an example case:

```python
from scipy import sparse
import numpy as np

csr1 = sparse.csr_array(sparse.random(10, 10, density=0.25, format="csr"))
csr2 = sparse.csr_array(sparse.random(10, 10, density=0.25, format="csr"))
csr1.indices = csr1.indices.astype(np.int64)

result = sparse.vstack([csr1, csr2])
result.indices.dtype
```

What should the resulting index dtype be? I *think* it should be `int64` following type promotion rules.

The alternative is to downcast to the smallest possible size, but this is the kind of behavior we removed.